Invariant cleanup goto programs jr

hannes-steffenhagen-diffblue opened the pull request:
Only 793fe3b and later are new (will require rebase later, this is just up for review now).

hannes-steffenhagen-diffblue committed:
Cleanup invariants in lazy_goto_model

hannes-steffenhagen-diffblue committed:
Cleanup invariants in link_goto_model

hannes-steffenhagen-diffblue committed:
Cleanup invariants in loop_ids

hannes-steffenhagen-diffblue committed:
Cleanup invariants in osx_fat_reader

hannes-steffenhagen-diffblue committed:
Cleanup invariants in parameter_assignments

hannes-steffenhagen-diffblue committed:
Cleanup invariants in pointer_arithmetic

hannes-steffenhagen-diffblue committed:
Cleanup invariants in printf_formatter

hannes-steffenhagen-diffblue committed:
Cleanup invariants in read_bin_goto_object

hannes-steffenhagen-diffblue committed:
Cleanup invariants in remove_complex

hannes-steffenhagen-diffblue committed:
Cleanup invariants in remove_const_function_pointers

hannes-steffenhagen-diffblue committed:
Cleanup invariants in remove_function_pointers

hannes-steffenhagen-diffblue committed:
Cleanup invariants in remove_unused_functions

hannes-steffenhagen-diffblue committed:
Cleanup invariants in remove_vector

hannes-steffenhagen-diffblue committed:
Cleanup invariants in replace_calls

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 4f9e5c83f30d9669a51d2594bd898a77f9e23fb1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85033090

peterschrammel commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -135,8 +134,10 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
-        throw 0;
+        throw deserialization_exceptiont(
+          "failed to figure out type of file"
+          "\nsource location: " +
```
A structured variant would be good to enable proper JSON/XML output of the source location.

hannes-steffenhagen-diffblue replied:
Actually, it seems like the source location here is only used to communicate the filename. I'll reword the error message.

hannes-steffenhagen-diffblue replied:
In this case the source location is actually just used to communicate the file name, so I think I'll just reword the message

peterschrammel commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -94,8 +95,7 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
   {
-    msg.error() << "Please provide a program" << messaget::eom;
-    throw 0;
+    throw invalid_user_input_exceptiont("No program provided", "", "");
```
Provide a proper constructor for such a case

peterschrammel commented on src/goto-programs/link_goto_model.cpp:
```diff
@@ -171,15 +172,19 @@ void link_goto_model(
-
+  {
+    // TODO more helpful error message
+    throw deserialization_exceptiont("Typechecking main failed");
```
start with lower case (several occurrences)

peterschrammel commented on src/goto-programs/remove_vector.cpp:
```diff
@@ -114,13 +115,14 @@ static void remove_vector(exprt &expr)
     else if(expr.id()==ID_unary_minus || expr.id()==ID_bitnot)
     {
+      DATA_INVARIANT(
+        expr.operands().size() == 1, "unary operators have one operands");
```
operand

peterschrammel commented on src/goto-programs/replace_calls.cpp:
```diff
@@ -105,8 +106,9 @@ void replace_callst::operator()(
-          throw "Returns must not be removed before replacing calls";
+        INVARIANT(
+          !has_suffix(id2string(se.get_identifier()), RETURN_VALUE_SUFFIX),
+          "this happens before returns are removed");
```
The original message was somehow better.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -158,8 +159,8 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
     {
-      msg.error() << "CONVERSION ERROR" << messaget::eom;
-      throw 0;
+      // TODO more helpful error message
```
I'm not sure this is a TODO (and neither in the parser case): those "more helpful error messages" are provided already, during parsing/type checking.

tautschnig commented on src/goto-programs/osx_fat_reader.cpp:
```diff
@@ -13,6 +13,8 @@ Module: Read Mach-O
 #include <cassert>
 #include <cstdlib>
+#include <util/exception_utils.h>
+#include <util/invariant.h>
```
`cassert` should no longer be necessary.

tautschnig commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -32,10 +32,10 @@ static exprt complex_member(const exprt &expr, irep_idt id)
   else
   {
-    assert(expr.type().id()==ID_struct);
+    PRECONDITION(expr.type().id() == ID_struct);
```
That's actually checked by the subsequent `to_struct_type`

tautschnig commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -222,12 +229,12 @@ static void remove_complex(exprt &expr)
   if(expr.id()==ID_complex_real)
   {
-    assert(expr.operands().size()==1);
+    static_cast<void>(to_complex_real_expr(expr));
```
Better use this in the subsequent statement: `expr = complex_member(to_complex_real_expr(expr).op(), ID_real)`

tautschnig commented on src/goto-programs/remove_unused_functions.cpp:
```diff
@@ -78,7 +78,10 @@ void find_used_functions(
+          INVARIANT(
+            call.function().id() == ID_symbol,
+            "by this point, all kinds of function calls should have been "
+            "converted to simple function calls");
```
Maybe add the error message as a comment in here, but really `to_symbol_expr` below actually takes care of this.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 5ec6f927340837b7e416167305a42f3e7609258d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86723412

hannes-steffenhagen-diffblue commented:
@peterschrammel @tautschnig I believe I've addressed your comments on here now, could you please re-review?

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -17,6 +17,7 @@
 #include <langapi/language.h>
 
 #include <fstream>
+#include <util/exception_utils.h>
```
Nit pick: please move this a few lines up, with the other `util` includes.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -122,21 +126,17 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
-                    << '\'' << messaget::eom;
-        throw 0;
+        throw system_exceptiont(
+          "failed to open input file `" + filename + '\'');
```
Isn't this (most likely) also an `invalid_user_input_exceptiont`. It's of course possible that your file system is somehow hosed, but in almost all cases the user will be the one  who got it wrong.

hannes-steffenhagen-diffblue replied:
It's quite possible that the root cause is because the user passed in the wrong filename. Or it could fail for other reasons. The only thing we know at this point is that we can't open the file, so we report that as such here.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -147,8 +147,7 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
       {
-        msg.error() << "PARSING ERROR" << messaget::eom;
-        throw 0;
+        throw deserialization_exceptiont("PARSING ERROR");
```
Isn't this invalid input? (I.e., another case of a user error.)

hannes-steffenhagen-diffblue replied:
`deserialization_exceptiont` is for "invalid inputs", but of another kind. If you look at `invalid_user_input_exceptiont` it's actually just indicating malformed arguments (IMHO it should be renamed to something like `bad_argument_exceptiont`), whereas `deserialization_exceptiont` means we got some input but it wasn't in the format that we expected.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -158,8 +157,7 @@ void lazy_goto_modelt::initialize(const cmdlinet &cmdline)
     {
-      msg.error() << "CONVERSION ERROR" << messaget::eom;
-      throw 0;
+      throw deserialization_exceptiont("CONVERSION ERROR");
```
As above: I don't think `deserialization_exceptiont` is the right choice.

tautschnig commented on src/goto-programs/link_goto_model.cpp:
```diff
@@ -171,15 +172,19 @@ void link_goto_model(
-      linking.object_type_updates))
-    throw 0;
+  {
+    throw deserialization_exceptiont("typechecking main failed");
```
As above: this is the result of invalid user input.

tautschnig commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -105,20 +104,22 @@ static void remove_complex(exprt &expr)
-          ID_complex_real, expr.op0(), expr.op0().type().subtype());
+          ID_complex_real,
+          typecast_expr.op(),
+          typecast_expr.op().type().subtype());
```
We now have a `complex_real_exprt`, please use this one in here while changing those lines.

tautschnig commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -153,62 +155,66 @@ static void remove_complex(exprt &expr)
     {
-      assert(expr.operands().size()==2);
+      static_cast<void>(to_complex_expr(expr));
       expr.id(ID_struct);
```
Maybe we should just make this code a bit more explicit:
```
struct_exprt s(expr.type());
s.add_source_location() = expr.source_location();
s.copy_to_operands(to_complex_expr(expr).real(), to_complex_expr(expr).imag());
expr.swap(s);
```

hannes-steffenhagen-diffblue replied:
I've been meaning to ask, why do we prefer swap over assignment in these cases?

tautschnig replied:
Swapping two `irept` is literally just swapping two pointers, doing `operator=` involves a sequence of reference-count updates -- unless rvalue references can be used, at which point it's just a swap again. But then the mentioned reference-count updates may happen either way if one of the swapped `irept`s goes out of scope. So it's probably an unmeasurable difference.

tautschnig commented on src/goto-programs/remove_vector.cpp:
```diff
@@ -108,13 +109,14 @@ static void remove_vector(exprt &expr)
     else if(expr.id()==ID_unary_minus || expr.id()==ID_bitnot)
     {
+      DATA_INVARIANT(
+        expr.operands().size() == 1, "unary operators have one operand");
```
Could still use `to_unary_expr` instead

tautschnig commented on src/goto-programs/remove_vector.cpp:
```diff
@@ -82,13 +82,14 @@ static void remove_vector(exprt &expr)
        expr.id()==ID_bitand || expr.id()==ID_bitor)
     {
+      DATA_INVARIANT(
+        expr.operands().size() == 2, "binary operators have 2 operands");
```
How about `to_binary_expr`?

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a7c8bab9b77214995a47871f0350197f7f36ad69).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86762862

hannes-steffenhagen-diffblue commented:
I'll use the `invalid_source_file_exceptiont` for this, but it's blocked on https://github.com/diffblue/cbmc/pull/3099 at the moment (travis taking very long for some reason).

tautschnig commented:
Should be unblocked by now (#3099 has been merged).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 3fa5d575c1b2f83e1f41e1ad76a2abc4250f49d4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87482764
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: e9067f459b73de396799d107fe7bd52c0f1cf01e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87486188

hannes-steffenhagen-diffblue commented:
@tautschnig I've changed the exception types (except for one system_exceptiont where I still think it's the right one to use, even though I see the argument). Would you mind taking another look?

martin-cs commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -131,7 +129,8 @@ static void remove_complex(exprt &expr)
     {
-      assert(expr.operands().size()==2);
+      DATA_INVARIANT(
+        expr.operands().size() == 2, "binary operators have 2 arguments");
```
But plus and mult are not binary:

https://github.com/diffblue/cbmc/blob/develop/src/util/std_expr.h#L963
https://github.com/diffblue/cbmc/blob/develop/src/util/std_expr.h#L1080

I realise it may be beyond the remit of this PR to fix this but this needs to be a PRECONDITION stating clearly that _in_this_context_ they must have 2 arguments.

hannes-steffenhagen-diffblue replied:
I added a comment explaining that this is unusual, and possibly an error.

martin-cs replied:
That's not strictly correct.  ID_plus, ID_mult and the other associative operators below are really genuinely, always not binary operators in the sense they do not inherit from binary_exprt.  However, in common operation, they often only have two operands.

There used to be some confusion over whether they were binary or n-arty so there is some code that is written with an implicit assumption that they are binary.  This and remove_vector are examples.  The ambiguity was resolved in favour of them becoming n-ary so this code is wrong.

It is indeed a DATA_INVARIANT that binary operators have two arguments but these are not binary operators so I would recommend a PRECONDITION instead, along with a FIXME saying it needs to handle the case of n-ary plus and mult.  Similar should be applied below.

martin-cs commented on src/goto-programs/remove_unused_functions.cpp:
```diff
@@ -77,9 +77,6 @@ void find_used_functions(
 
-          // check that this is actually a simple call
-          assert(call.function().id()==ID_symbol);
-
```
Where has this check gone?

hannes-steffenhagen-diffblue replied:
There is a `to_symbol_expr` below this line that does the assert. @tautschnig preferred to do this check only once and I had no preference.

martin-cs replied:
Fair.  He and you are correct.

martin-cs commented on src/goto-programs/remove_vector.cpp:
```diff
@@ -82,13 +82,13 @@ static void remove_vector(exprt &expr)
        expr.id()==ID_mod  || expr.id()==ID_bitxor ||
        expr.id()==ID_bitand || expr.id()==ID_bitor)
     {
+      auto const &binary_expr = to_binary_expr(expr);
```
bitand is not binary:

https://github.com/diffblue/cbmc/blob/develop/src/util/std_expr.h#L2716

hannes-steffenhagen-diffblue replied:
I added a comment pointing out that this is an unusual construct.

martin-cs reviewed (changes requested):
The older code is confused about which are binary and which are n-ary operations.  This is not uncommon as I remember submitting the patch that started clarifying this.  My guess is that there aren't tests for this.  Fixing it may be beyond the scope of this PR but we should at least have the checks and invariants correctly annotated.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 4a46e0964f23aec4b26c590ccb1a89b1a27d0237).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/87670052

hannes-steffenhagen-diffblue commented:
@martin-cs I changed the DATA_INVARIANT to precondition, added a precondition in the other case where it just used to_binary_expr without a check before, and reworded the fixme comment to make it more clear that the operations are n-ary operations specifically rather than just "not binary".

martin-cs commented:
@hannes-steffenhagen-diffblue : Thank you.

martin-cs reviewed (approved):
All issues I raised have been handled appropriately.

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -170,7 +168,12 @@ void lazy_goto_modelt::initialize(
-      throw 0;
+    {
+      source_locationt source_location;
+      source_location.set_file(irep_idt(file));
```
I am surprised you actually need the `irep_idt` in there?

hannes-steffenhagen-diffblue replied:
We don't. Not sure why that's in here, I'll remove it.

tautschnig commented on src/goto-programs/remove_complex.cpp:
```diff
@@ -153,62 +155,69 @@ static void remove_complex(exprt &expr)
-      assert(expr.operands().size()==2);
-      expr.id(ID_struct);
+      auto const &complex_expr = to_complex_expr(expr);
+      struct_exprt struct_expr{complex_expr.type()};
```
Nit pick: use round parentheses

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 75414ce278cccb7a4b04a36b9eba059e87a5d228).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88091760

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 365bb722b5234826bfff7dde2fdbe4dce8083f3b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/88137222
